File: okba_challenge1/c/sol.c
#include<stdio.h>
#include <string.h>
/*
example input :
1@#AAibddwncPhOlaOinoqa5
output :
AAOOddhlP15@#iaioabwncnq
*/
int main()
{
    // read input
    char s[256];
    scanf("%s", &s);

    // solution
    char capitalVowels[256] = "";
    char smallDiv4[256] = "";
    char capitalContstantes[256] = "";
    char capitalDiv4[256] = "";
    char digits[256] = "";
    char symboles[256] = "";
    char smallVowels[256] = "";
    char smallConstantes[256] = "";

    for (int i = 0; i < strlen(s); i++) {
        if (strchr("AEIOUY",s[i]))
                strncat(capitalVowels, &s[i], 1);
            else if (strchr("dhlptx",s[i]))
                strncat(smallDiv4, &s[i], 1);
            else if (strchr("BCDFGHJKLMNOPQRSTVWXZ",s[i]))
                strncat(capitalContstantes, &s[i], 1);
            else if (strchr("DHLPTX",s[i]))
                strncat(capitalDiv4, &s[i], 1);
            else if (strchr("0123456789",s[i]))
                strncat(digits, &s[i], 1);
            else if (strchr("!@#$%^&*()_+/\\][{}\":;`<>?\'",s[i]))
                strncat(symboles, &s[i], 1);
            else if (strchr("aeiouy",s[i]))
                strncat(smallVowels, &s[i], 1);
            else if (strchr("bcdfghjklmnpqrstvwxz",s[i]))
                strncat(smallConstantes, &s[i], 1);

    }

    printf("%s%s%s%s%s%s%s%s", capitalVowels, smallDiv4, capitalContstantes, capitalDiv4, digits, symboles, smallVowels, smallConstantes);
    return 0;
}

File: okba_challenge1/Sol.pas
program OptimalSort;

var
  capitalVowels, smallDiv4, capitalConsonants, capitalDiv4, digits, symbols, smallVowels, smallConsonants, myString: String;
  i: Integer;
  ch: Char;

begin
  capitalVowels := '';
  smallDiv4 := '';
  capitalConsonants := '';
  capitalDiv4 := '';
  digits := '';
  symbols := '';
  smallVowels := '';
  smallConsonants := '';

  Write('Enter the string to sort: ');
  Readln(myString);

  for i := 1 to Length(myString) do
  begin
    ch := myString[i];
    if Pos(ch, 'AEIOUY') <> 0 then
      capitalVowels := capitalVowels + ch
    else if Pos(ch, 'dhlptx') <> 0 then
      smallDiv4 := smallDiv4 + ch
    else if Pos(ch, 'BCDFGHJKLMNOPQRSTVWXZ') <> 0 then
      capitalConsonants := capitalConsonants + ch
    else if Pos(ch, 'DHLPTX') <> 0 then
      capitalDiv4 := capitalDiv4 + ch
    else if Pos(ch, '0123456789') <> 0 then
      digits := digits + ch
    else if Pos(ch, '!@#$%^&*()_+/\][{}":;`<>?''') <> 0 then
      symbols := symbols + ch
    else if Pos(ch, 'aeiouy') <> 0 then
      smallVowels := smallVowels + ch
    else if Pos(ch, 'bcdfghjklmnpqrstvwxz') <> 0 then
      smallConsonants := smallConsonants + ch;
  end;

  Writeln(capitalVowels + smallDiv4 + capitalConsonants + capitalDiv4 + digits + symbols + smallVowels + smallConsonants);
	Readln;
end.
